### PR TITLE
Rendering system

### PR DESCRIPTION
**Pull request by z-adams**
My progress so far on the new rendering system. It doesn't yet support offscreen rendering or complicated multipass, but it currently has feature parity with the existing rendering scheme, but is structured in a much better way.

Objects are given an `ACompVisible` component to indicate they should be processed by the renderer.

Shader instances no longer exist; the rendering system (or an external system, if a nonstandard shader is being used) is responsible for creating an `ACompShader` which stores a function pointer to a static shader draw function, which is now responsible for collecting any information necessary to set uniforms and draw the object.

### CODE REVIEW

File: bin/OSPData/adera/Shaders/FullscreenTri.frag [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+layout(location = 0, index = 0) out vec3 color;
+
+layout(location = 0) uniform sampler2D framebuffer;
+
+in vec2 uv;
+
+void main()
+{
+    color = texture(framebuffer, uv).rgb;
+}
```
Comment (jonesmz) [inline bin/OSPData/adera/Shaders/FullscreenTri.frag:37]: no newline at end of file?

makes a lot of editors complain.

File: src/adera/Shaders/PlumeShader.cpp [hunk 2]
```diff
@@ -24,31 +24,53 @@
...
+#include <osp/Active/SysRender.h>
+#include <osp/Resource/AssetImporter.h>
+
 #include <Magnum/GL/Version.h>
 #include <Magnum/GL/Shader.h>
 #include <Corrade/Containers/Reference.h>
 #include <Magnum/GL/Texture.h>
 
-using namespace Magnum;
+using Magnum::Int;
```
Comment (jonesmz) [inline src/adera/Shaders/PlumeShader.cpp:36]: ugly ;)

File: src/adera/Shaders/PlumeShader.cpp [hunk 3]
```diff
@@ -24,31 +24,53 @@
...
+
+    using Magnum::GL::Texture2D;
+    using Magnum::GL::Mesh;
+    // Collect uniform data
+    auto& transform = rScene.reg_get<ACompTransform>(e);
+    auto& rPlumeComp = rScene.reg_get<ACompExhaustPlume>(e);
+    auto const& plumedata = *rPlumeComp.m_effect;
+    Mesh& rMesh = *rScene.reg_get<ACompMesh>(e).m_mesh;
+
+    constexpr std::string_view texName = "OSPData/adera/noise1024.png";
```
Comment (jonesmz) [inline src/adera/Shaders/PlumeShader.cpp:56]: no need to change it for this PR, but eventually we should make sure either:
1) Textures like this are directly imported into the binary, there are various techniques for how to do this, so i won't detail the mechanism that's used, but the point is that it's not loaded from disk.
2) Resources loaded from disk are described in a config file, so that they can be changed by the end user.

File: src/osp/Active/ActiveScene.cpp [hunk 4]
```diff
@@ -216,25 +216,33 @@ void ActiveScene::update_hierarchy_transforms()
...
-    //Matrix4 cameraInverse;
-
-    // TODO: check if camera has the right components
-    ACompCamera& cameraComp = reg_get<ACompCamera>(camera);
-    ACompTransform& cameraTransform = reg_get<ACompTransform>(camera);
-
-    //cameraProject = cameraComp.m_projection;
-    cameraComp.m_inverse = cameraTransform.m_transformWorld.inverted();
+    auto renderView = m_registry.view<ACompRenderingAgent, ACompPerspective3DView, ACompRenderer>();
+    for (auto [e, agent, view, renderer] : renderView.each())
```
Comment (jonesmz) [inline src/osp/Active/ActiveScene.cpp:226]: imho, this would be better as `auto const&`, but whether that actually results in different code gen depends on the specific types of the variables.

File: src/osp/Active/SysRender.cpp [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+
+        std::array<float, 12> surfData
+        {
+            // Vert position    // UV coordinate
+            -1.0f,  1.0f,       0.0f,  1.0f,
+            -1.0f, -3.0f,       0.0f, -1.0f,
+             3.0f,  1.0f,       2.0f,  1.0f
+        };
+
+        GL::Buffer surface(std::move(surfData), GL::BufferUsage::StaticDraw);
```
Comment (jonesmz) [inline src/osp/Active/SysRender.cpp:91]: std::array can't be moved from. It's equivelent to a raw c-style array on the stack.

As a result, you should probably make surfData a static constexpr variable, to give the compiler a better chance at not fucking up.

You could go a step further and make it an std::initializer_list<float> as well.

File: src/adera/Shaders/PlumeShader.cpp (hunk 3 above)
Comment (z-adams) [reply src/adera/Shaders/PlumeShader.cpp:56]: This will all be config'd eventually, although importing into the binary is an interesting idea for loading shaders.

File: src/osp/Active/SysRender.cpp [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        resources.add<GL::Framebuffer>("offscreen_fbo", std::move(fbo));
+
+    rScene.reg_emplace<ACompRenderTarget>(rScene.hier_get_root(),
+        viewSize, std::move(fboRes));
+    rScene.reg_emplace<ACompFBOColorAttachment>(rScene.hier_get_root(), colorRes);
+
+
+    // Create the default render pipeline
+    resources.add<RenderPipeline>("default",
+        std::move(create_render_pipeline()));
```
Comment (jonesmz) [inline src/osp/Active/SysRender.cpp:132]: if create_render_pipeline returns an object by value, or by const&, then this std::move() will have no effect.

Functions that return an object by value will automatically have the returned value converted to && in this context.

std::move() silently ignores const& parameters, returning a const&.

So the only time this std::move() will have an effect on the program is if create_render_pipeline() returns a value by non-const reference, and while that's not inherently problematic, given the name of the function, i'd argue that's not what this function should return.

So, long story short, the std::move() here can be removed.

Comment (z-adams) [reply src/osp/Active/SysRender.cpp:132]: Good to know, thanks

File: src/osp/Active/SysRender.cpp [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+            Renderer::setFaceCullingMode(Renderer::PolygonFacing::Front);
+            SysRender::draw_group(rScene, transparentView, camera);
+
+            // Draw frontfaces
+            Renderer::setFaceCullingMode(Renderer::PolygonFacing::Back);
+            SysRender::draw_group(rScene, transparentView, camera);
+        }
+    );
+
+    return {pipeline};
```
Comment (jonesmz) [inline src/osp/Active/SysRender.cpp:191]: I recommend returning a braced initializer list instead of giving the vector a name.


Uhhhhhh, also, why aren't you using the vector of handles?

File: src/osp/Active/SysRender.cpp [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+{
+    auto view = rScene.get_registry().view<ACompVisible>();
+    for (ActiveEnt e : view)
+    {
+        auto* shader = rScene.reg_try_get<ACompShader>(e);
+
+        if (shader == nullptr)
+        {
+            // Drawable entity has not been initialized yet
+            constexpr ShaderDrawFnc_t defaultDraw = shader::Phong::draw_entity;
```
Comment (jonesmz) [inline src/osp/Active/SysRender.cpp:204]: imho not really useful to have this defaultDraw variable.

File: src/osp/Active/SysRender.cpp (hunk 7 above)
Comment (z-adams) [reply src/osp/Active/SysRender.cpp:191]: I had it there originally, but it causes a crash when the resource is deleted at program exit, and I have no actual use for the handles (the only way to create a FunctionOrder is via constructing FunctionOrderHandles) so I discarded them.

File: src/osp/Active/SysRender.cpp [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+            shader = &rScene.reg_emplace<ACompShader>(e, defaultDraw);
+            rScene.reg_emplace<ACompOpaque>(e);
+        }
+    }
+
+    auto agentView = rScene.get_registry().view<ACompRenderingAgent>(entt::exclude<ACompRenderer>);
+    for (ActiveEnt e : agentView)
+    {
+        constexpr std::string_view defaultRenderer = "default";
+        rScene.reg_emplace<ACompRenderer>(e, std::string{defaultRenderer});
```
Comment (jonesmz) [inline src/osp/Active/SysRender.cpp:214]: I created the SharedString class specifically for write-once, read-many strings like this.

File: src/osp/Active/SysRender.cpp (hunk 8 above)
Comment (z-adams) [reply src/osp/Active/SysRender.cpp:204]: I separated it out like that because I'm not sure if I'm going to just hard-code it, or if it should be fetched from elsewhere.

File: src/osp/Active/SysRender.cpp (hunk 9 above)
Comment (z-adams) [reply src/osp/Active/SysRender.cpp:214]: I'll look into it, but I think ultimately these strings should probably be replaced by compile-time hashed keys that are easier to work with than strings.

File: src/osp/Active/SysVehicle.cpp [hunk 10]
```diff
@@ -366,20 +365,11 @@ std::pair<ActiveEnt, std::vector<SysVehicle::MachineDef>> SysVehicle::part_insta
...
-            rScene.reg_emplace<Phong::ACompPhongInstance>(currentEnt, std::move(shader));
-
-            CompDrawableDebug& bBocks = rScene.reg_emplace<CompDrawableDebug>(
-                        currentEnt, meshRes, &Phong::draw_entity, 0x0202EE_rgbf);
-
-            //new DrawablePhongColored(*obj, *m_shader, *mesh, 0xff0000_rgbf, m_drawables);
+            // by now, the mesh and texture should both exist so we emplace them
+            // as components to be consumed by the Phong shader
+            rScene.reg_emplace<ACompMesh>(currentEnt, meshRes);
+            rScene.reg_emplace<ACompDiffuseTex>(currentEnt, std::move(textureResources[0]));
```
Comment (jonesmz) [inline src/osp/Active/SysVehicle.cpp:371]: you move only the first texture? Why bother creating a vector if you are only ever using the first item in it?

File: src/osp/Shaders/FullscreenTriShader.cpp [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+{
+    set_framebuffer(texture);
+    draw(surface);
+}
+
+FullscreenTriShader& FullscreenTriShader::set_framebuffer(GL::Texture2D& rTex)
+{
+    rTex.bind(static_cast<Int>(TextureSlot::FramebufferSlot));
+    return *this;
+}
```
Comment (jonesmz) [inline src/osp/Shaders/FullscreenTriShader.cpp:59]: missing end of file newline

File: src/osp/Shaders/FullscreenTriShader.h [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+{
+public:
+    // Vertex attribs
+    typedef Magnum::GL::Attribute<0, Magnum::Vector2> Position;
+    typedef Magnum::GL::Attribute<1, Magnum::Vector2> TextureCoordinates;
+
+    // Outputs
+    enum : Magnum::UnsignedInt
+    {
+        ColorOutput = 0
```
Comment (jonesmz) [inline src/osp/Shaders/FullscreenTriShader.h:43]: Interesting trick,
 but enums that aren't enum classes can be implicitly converted to other types, can't they? so this doesn't buy you much.

File: src/osp/Active/SysVehicle.cpp (hunk 10 above)
Comment (z-adams) [reply src/osp/Active/SysVehicle.cpp:371]: Temporary ugliness. Eventually we'll be loading several textures, and the code above this is sort of sitting around waiting to be hammered into a proper PBR resource loader. It was like that before this PR.

File: src/osp/Shaders/FullscreenTriShader.h [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    // Uniforms
+    enum class UniformPos : Magnum::Int
+    {
+        FramebufferSampler = 0
+    };
+
+    // Texture2D slots
+    enum class TextureSlot : Magnum::Int
+    {
+        FramebufferSlot = 0
```
Comment (jonesmz) [inline src/osp/Shaders/FullscreenTriShader.h:65]: Since you always have to refer to this as TextureSlot::FramebufferSlot, maybe drop the trailing Slot suffix?

File: src/osp/Shaders/FullscreenTriShader.h (hunk 12 above)
Comment (z-adams) [reply src/osp/Shaders/FullscreenTriShader.h:43]: I'm just following Magnum's convention, this is how they suggest designing shader program wrappers, so I've stuck with it for the sake of consistency even though it's a bit verbose.

File: src/osp/Active/SysRender.cpp [hunk 14]
```diff
@@ -0,0 +1,275 @@
...
+
+            // Draw backfaces
+            Renderer::setFaceCullingMode(Renderer::PolygonFacing::Front);
+            SysRender::draw_group(rScene, transparentView, camera);
+
+            // Draw frontfaces
+            Renderer::setFaceCullingMode(Renderer::PolygonFacing::Back);
+            SysRender::draw_group(rScene, transparentView, camera);
+        }
+    );
```
Comment (Capital-Asterisk) [inline src/osp/Active/SysRender.cpp:189]: So these two render passes is now the "debug renderer"

File: src/osp/Active/SysRender.h [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+template<typename VIEW_T>
+void SysRender::draw_group(ActiveScene& rScene, VIEW_T& drawlist, ACompCamera const& camera)
+{
+    for (ActiveEnt e : drawlist)
+    {
+        auto shader = drawlist.template get<ACompShader>(e);
+
+        shader.m_drawCall(e, rScene, camera);
+    }
+}
```
Comment (Capital-Asterisk) [inline src/osp/Active/SysRender.h:162]: For stuff like filling a g-buffer or common objects, there should be a "draw group using this shader only" function too.

**Review by Capital-Asterisk — COMMENTED** [review]
A step in the right direction. Still expecting lots of stuff discussed in the Discord server.

File: src/adera/SysExhaustPlume.h [hunk 16]
```diff
@@ -35,6 +35,8 @@ struct ACompExhaustPlume
 {
     ActiveEnt m_parentMachineRocket{entt::null};
     DependRes<PlumeEffectData> m_effect;
+    float m_time{0.0f};
+    float m_powerLevel{0.0f};
```
[the remark this answers is not in the document]
Comment (z-adams) [reply src/adera/SysExhaustPlume.h:39]: PlumeEffectData was supposed to represent a plume config file, whose fields are just read straight into a struct. This will probably all shift around as we figure out what sorts of properties are needed to define a plume, or if multiple shaders are needed, etc. I put the time in there since before it was stored in the shader instance, although I suppose once ActiveScene provides a way to get the global time we could get rid of it.

File: src/osp/Active/SysRender.cpp (hunk 14 above)
Comment (z-adams) [reply src/osp/Active/SysRender.cpp:189]: Yep, or default. Ultimately it probably won't end up looking much different from this.

File: src/osp/Active/SysRender.h (hunk 15 above)
Comment (z-adams) [reply src/osp/Active/SysRender.h:162]: That would usually be handled as one of the render passes. Can you provide an example?

File: src/osp/Active/SysRender.cpp (hunk 7 above)
Comment (jonesmz) [reply src/osp/Active/SysRender.cpp:191]: Ah. That's terrible :-)

File: src/osp/Active/SysVehicle.cpp (hunk 10 above)
Comment (jonesmz) [reply src/osp/Active/SysVehicle.cpp:371]: understood

File: src/osp/Shaders/FullscreenTriShader.h (hunk 12 above)
Comment (jonesmz) [reply src/osp/Shaders/FullscreenTriShader.h:43]: Understood.

Kind of feel like you should promote this to a proper enum class, but you do you.

Comment (z-adams) [reply src/osp/Shaders/FullscreenTriShader.h:43]: Oh I thought it was an enum class, that's just a typo. The Magnum way uses enum classes and I have everywhere except apparently here. I'll fix it.

File: src/osp/Active/ActiveScene.cpp (hunk 4 above)
Comment (z-adams) [reply src/osp/Active/ActiveScene.cpp:226]: This doesn't really seem to change the types that get used, which are `T&&` for the entity type (`e`) and `T&` for all the component types. This doesn't change even if I add `const`. I'm going to leave it as-is since this is the way skypjack himself seems to write it.

File: src/osp/Active/SysRender.cpp [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+    /* Configure pipeline resources */
+
+    // Add an offscreen framebuffer
+
+    Vector2i viewSize = GL::defaultFramebuffer.viewport().size();
+
+    GL::Texture2D color;
+    color.setStorage(1, GL::TextureFormat::RGB8, viewSize);
+    DependRes<GL::Texture2D> colorRes =
+        resources.add<GL::Texture2D>("offscreen_fbo_color", std::move(color));
```
Comment (Capital-Asterisk) [inline src/osp/Active/SysRender.cpp:111]: Package::add is actually an emplace, you can do this:
```cpp
DependRes<GL::Texture2D> colorRes =
        resources.add<GL::Texture2D>("offscreen_fbo_color");
colorRes->setStorage(1, GL::TextureFormat::RGB8, viewSize);
```

File: src/osp/Active/SysRender.cpp [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+
+    auto agentView = rScene.get_registry().view<ACompRenderingAgent>(entt::exclude<ACompRenderer>);
+    for (ActiveEnt e : agentView)
+    {
+        constexpr std::string_view defaultRenderer = "default";
+        rScene.reg_emplace<ACompRenderer>(e, std::string{defaultRenderer});
+    }
+}
+
+void SysRender::render(ActiveScene& rScene, ACompCamera const& camera)
```
Comment (Capital-Asterisk) [inline src/osp/Active/SysRender.cpp:217]: Since render order is no longer being called, this function is no longer being called and is leftover?

File: src/osp/Active/SysRender.cpp [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+        auto* shader = rScene.reg_try_get<ACompShader>(e);
+
+        if (shader == nullptr)
+        {
+            // Drawable entity has not been initialized yet
+            constexpr ShaderDrawFnc_t defaultDraw = shader::Phong::draw_entity;
+            shader = &rScene.reg_emplace<ACompShader>(e, defaultDraw);
+            rScene.reg_emplace<ACompOpaque>(e);
+        }
+    }
```
Comment (Capital-Asterisk) [inline src/osp/Active/SysRender.cpp:207]: This is a lot sketchier than I imagined. Since ACompShader just stores a function pointer now, it's safe to set the draw function on initialization.

File: src/osp/Active/SysRender.cpp (hunk 18 above)
Comment (z-adams) [reply src/osp/Active/SysRender.cpp:217]: Yep, just forgot to remove it

File: src/osp/Active/SysRender.cpp (hunk 19 above)
Comment (z-adams) [reply src/osp/Active/SysRender.cpp:207]: Elaborate?

Comment (Capital-Asterisk) [reply src/osp/Active/SysRender.cpp:207]: `shader::Phong::draw_entity` can be stored somewhere in the scene or package as the "default shader". SysVehicle can use this this to make an ACompShader.

**Review by Capital-Asterisk — APPROVED** [review]

File: src/test_application/flight.cpp [hunk 20]
```diff
@@ -121,15 +134,18 @@ void testapp::test_flight(std::unique_ptr<OSPMagnum>& pMagnumApp,
...
 
     // Create the camera entity
     ActiveEnt camera = rScene.hier_create_child(rScene.hier_get_root(),
                                                        "Camera");
     auto &cameraTransform = rScene.reg_emplace<ACompTransform>(camera);
-    auto &cameraComp = rScene.get_registry().emplace<ACompCamera>(camera);
+    auto &cameraComp = rScene.reg_emplace<ACompCamera>(camera);
+    rScene.reg_emplace<ACompRenderingAgent>(camera, SysRender::get_default_rendertarget(rScene));
+    rScene.reg_emplace<ACompPerspective3DView>(camera, camera);
+    rScene.reg_emplace<ACompRenderer>(camera);
```
Comment (Capital-Asterisk) [inline src/test_application/flight.cpp:148]: It would really help to organize the setup in one readable section of code, such as:
```cpp
    // Create the camera entity
    ActiveEnt camera = rScene.hier_create_child(rScene.hier_get_root(),
                                                "Camera");
    auto &cameraTransform = rScene.reg_emplace<ACompTransform>(camera);
    auto &cameraComp = rScene.reg_emplace<ACompCamera>(camera);
    rScene.reg_emplace<ACompFloatingOrigin>(camera);
    cameraTransform.m_transform = Matrix4::translation(Vector3(0, 0, 25));
    cameraComp.m_far = 1u << 24;
    cameraComp.m_near = 1.0f;
    cameraComp.m_fov = 45.0_degf;
    cameraComp.calculate_projection();
    cameraComp.m_viewport
            = Vector2(Magnum::GL::defaultFramebuffer.viewport().size());

    // Add the debug camera controller to the scene. This adds controls
    auto camObj = std::make_unique<DebugCameraController>(rScene, camera);

    // Add a ACompDebugObject to camera to manage camObj's lifetime
    rScene.reg_emplace<ACompDebugObject>(camera, std::move(camObj));

    // Setup default render system
    osp::active::SysRender::setup(rScene);
    load_shaders(rScene);
    
    // Setup scene renderer, added to the camera component
    rScene.reg_emplace<ACompRenderingAgent>(camera, SysRender::get_default_rendertarget(rScene));
    rScene.reg_emplace<ACompPerspective3DView>(camera, camera);
    rScene.reg_emplace<ACompRenderer>(camera);
```

Comment (z-adams) [reply src/test_application/flight.cpp:148]: Sounds good, I will make that change and then merge
